File: 项目文件/frontend/src/api/systems.ts
import { request } from '../utils/request';
import type {
  SystemFormValues,
  SystemListResponse,
  SystemRecord,
} from '../types/system';
import type { UnifiedResponse } from '../types/user';

/**
 * 后端 SystemResponse.is_vm 是未序列化的计算属性（property），
 * API 响应不含该字段，前端基于 parent_system_id 非空即 VM 归一化计算。
 */
function withIsVm(record: Omit<SystemRecord, 'is_vm'>): SystemRecord {
  return { ...record, is_vm: record.parent_system_id != null };
}

export type SystemCreateData = SystemFormValues & { server_id: string };

export async function listSystems(params?: {
  server_id?: string;
  parent_system_id?: string;
  page?: number;
  page_size?: number;
}): Promise<UnifiedResponse<SystemListResponse>> {
  const searchParams = new URLSearchParams();
  if (params?.server_id) searchParams.set('server_id', params.server_id);
  if (params?.parent_system_id) searchParams.set('parent_system_id', params.parent_system_id);
  if (params?.page) searchParams.set('page', String(params.page));
  if (params?.page_size) searchParams.set('page_size', String(params.page_size));
  const query = searchParams.toString();
  const res = await request<SystemListResponse>(`/systems/${query ? `?${query}` : ''}`);
  if (res.code === 0) {
    return { ...res, data: { ...res.data, items: res.data.items.map(withIsVm) } };
  }
  return res;
}

export async function createSystem(data: SystemCreateData): Promise<UnifiedResponse<SystemRecord>> {
  const res = await request<SystemRecord>('/systems/', { method: 'POST', body: data });
  if (res.code === 0) {
    return { ...res, data: withIsVm(res.data) };
  }
  return res;
}

export async function getSystem(id: string): Promise<UnifiedResponse<SystemRecord>> {
  const res = await request<SystemRecord>(`/systems/${id}/`);
  if (res.code === 0) {
    return { ...res, data: withIsVm(res.data) };
  }
  return res;
}

export async function updateSystem(
  id: string,
  data: Partial<SystemFormValues>
): Promise<UnifiedResponse<SystemRecord>> {
  const res = await request<SystemRecord>(`/systems/${id}/`, { method: 'PUT', body: data });
  if (res.code === 0) {
    return { ...res, data: withIsVm(res.data) };
  }
  return res;
}

export async function deleteSystem(id: string): Promise<UnifiedResponse<null>> {
  return request<null>(`/systems/${id}/`, { method: 'DELETE' });
}

File: 项目文件/backend/app/api/systems.py
"""系统管理 API 路由"""

import uuid

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.deps import get_current_user
from app.models.user import User
from app.schemas.common import UnifiedResponse
from app.schemas.system import (
    SystemCreate,
    SystemListResponse,
    SystemResponse,
    SystemUpdate,
)
from app.services.system import (
    create_system,
    delete_system,
    get_system,
    list_systems,
    update_system,
)

router = APIRouter()


@router.get("/", response_model=UnifiedResponse[SystemListResponse])
async def list_systems_endpoint(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    server_id: uuid.UUID | None = Query(None),
    parent_system_id: uuid.UUID | None = Query(None),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """查询系统列表，支持按 server_id / parent_system_id 过滤"""
    systems, total = await list_systems(
        db, current_user.id, server_id=server_id, parent_system_id=parent_system_id, page=page, page_size=page_size,
    )
    return UnifiedResponse(
        data=SystemListResponse(
            items=[SystemResponse.model_validate(s) for s in systems],
            total=total,
        )
    )


@router.post("/", response_model=UnifiedResponse[SystemResponse])
async def create_system_endpoint(
    request: SystemCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """创建系统"""
    system = await create_system(db, current_user.id, request)
    return UnifiedResponse(data=SystemResponse.model_validate(system))


@router.get("/{system_id}", response_model=UnifiedResponse[SystemResponse])
async def get_system_endpoint(
    system_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """获取系统详情"""
    system = await get_system(db, system_id, current_user.id)
    if not system:
        raise HTTPException(status_code=404, detail="系统不存在")
    return UnifiedResponse(data=SystemResponse.model_validate(system))


@router.put("/{system_id}", response_model=UnifiedResponse[SystemResponse])
async def update_system_endpoint(
    system_id: uuid.UUID,
    request: SystemUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """更新系统"""
    system = await update_system(db, system_id, current_user.id, request)
    if not system:
        raise HTTPException(status_code=404, detail="系统不存在")
    return UnifiedResponse(data=SystemResponse.model_validate(system))


@router.delete("/{system_id}", response_model=UnifiedResponse[None])
async def delete_system_endpoint(
    system_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """删除系统"""
    success = await delete_system(db, system_id, current_user.id)
    if not success:
        raise HTTPException(status_code=404, detail="系统不存在")
    return UnifiedResponse(msg="系统已删除")
